File: ios/RecipeClipper/UI/Clip/ClipWebPage.swift
import SwiftUI
import WebKit

/// What the page reports, decoded from `clipper.js`'s messages.
enum ClipPageEvent: Equatable {
    case selection(String)
    case tagTapped(ClipField)
    case imageTapped(String)

    static func decode(_ json: String) -> ClipPageEvent? {
        guard let data = json.data(using: .utf8),
              let message = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = message["type"] as? String
        else { return nil }
        switch type {
        case "selection": return .selection(message["text"] as? String ?? "")
        case "tag": return (message["field"] as? String).flatMap(ClipField.init(rawValue:)).map { .tagTapped($0) }
        case "image": return (message["src"] as? String).flatMap { $0.isEmpty ? nil : .imageTapped($0) }
        default: return nil
        }
    }
}

/// `shared/web/clipper.js`, bundled as the web/ folder (one copy for both platforms).
enum ClipperScript {
    static let source: String = {
        guard let url = Bundle.main.url(forResource: "clipper", withExtension: "js", subdirectory: "web"),
              let source = try? String(contentsOf: url, encoding: .utf8)
        else { fatalError("web/clipper.js is missing from the bundle") }
        return source
    }()
}

/// The page being clipped, in a `WKWebView` with `clipper.js` injected at document end. The view
/// layer's half of the bridge (Android's ClipWebPage): it forwards the page's events and pushes
/// `syncState` and `pickingPhoto` into the page whenever they change. Links to other pages are
/// blocked, so the clip always comes from the page it is saved under. `fixtureHTML` replaces the
/// live page in UI tests.
struct ClipWebPage: UIViewRepresentable {
    let url: String
    var fixtureHTML: String? = nil
    let syncState: String
    let pickingPhoto: Bool
    let onEvent: (ClipPageEvent) -> Void

    func makeCoordinator() -> Coordinator { Coordinator() }

    func makeUIView(context: Context) -> WKWebView {
        let coordinator = context.coordinator
        coordinator.onEvent = onEvent
        let config = WKWebViewConfiguration()
        config.userContentController.addUserScript(
            WKUserScript(source: ClipperScript.source, injectionTime: .atDocumentEnd, forMainFrameOnly: true)
        )
        config.userContentController.add(WeakMessageHandler(coordinator), name: "rc")
        let webView = WKWebView(frame: .zero, configuration: config)
        webView.navigationDelegate = coordinator
        webView.accessibilityIdentifier = "clip.page"
        if let fixtureHTML {
            webView.loadHTMLString(fixtureHTML, baseURL: URL(string: url))
        } else if let pageUrl = URL(string: url) {
            webView.load(URLRequest(url: pageUrl))
        }
        return webView
    }

    func updateUIView(_ webView: WKWebView, context: Context) {
        let coordinator = context.coordinator
        coordinator.onEvent = onEvent
        coordinator.script = "window.RC && (RC.sync(\(syncState)), RC.pickImage(\(pickingPhoto)));"
        coordinator.push(to: webView)
    }

    static func dismantleUIView(_ webView: WKWebView, coordinator: Coordinator) {
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "rc")
    }

    @MainActor
    final class Coordinator: NSObject, WKNavigationDelegate, WKScriptMessageHandler {
        var onEvent: (ClipPageEvent) -> Void = { _ in }
        var script = ""
        private var loaded = false
        private var sent: String?

        /// Sends the latest state once the page has the script, and only when it changed.
        func push(to webView: WKWebView) {
            guard loaded, script != sent else { return }
            sent = script
            webView.evaluateJavaScript(script, completionHandler: nil)
        }

        func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
            loaded = false
            sent = nil
        }

        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            loaded = true
            sent = nil
            push(to: webView)
        }

        /// Redirects, the first load and fragment jumps load; a tapped link to any other page,
        /// or a form sent to one, does not.
        func webView(
            _ webView: WKWebView,
            decidePolicyFor navigationAction: WKNavigationAction,
            decisionHandler: @escaping @MainActor (WKNavigationActionPolicy) -> Void
        ) {
            let userNavigation = navigationAction.navigationType == .linkActivated
                || navigationAction.navigationType == .formSubmitted
            guard navigationAction.targetFrame?.isMainFrame == true, userNavigation else {
                decisionHandler(.allow)
                return
            }
            decisionHandler(Self.samePage(navigationAction.request.url, webView.url) ? .allow : .cancel)
        }

        private static func samePage(_ a: URL?, _ b: URL?) -> Bool {
            guard let a, let b else { return false }
            var left = URLComponents(url: a, resolvingAgainstBaseURL: true)
            var right = URLComponents(url: b, resolvingAgainstBaseURL: true)
            left?.fragment = nil
            right?.fragment = nil
            return left?.url == right?.url
        }

        func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
            guard let json = message.body as? String, let event = ClipPageEvent.decode(json) else { return }
            onEvent(event)
        }
    }
}

/// The content controller holds its handlers strongly; this keeps it from holding the
/// coordinator (and through it the view) alive.
private final class WeakMessageHandler: NSObject, WKScriptMessageHandler {
    weak var target: WKScriptMessageHandler?
    init(_ target: WKScriptMessageHandler) { self.target = target }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        target?.userContentController(userContentController, didReceive: message)
    }
}

File: app/src/main/java/com/example/recipeclipper/ui/clip/ClipWebPage.kt
package com.example.recipeclipper.ui.clip

import android.annotation.SuppressLint
import android.net.Uri
import androidx.core.net.toUri
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView
import com.example.recipeclipper.data.model.ClipField
import org.json.JSONObject

/** What the page reports, decoded from `clipper.js`'s messages. */
internal sealed class ClipPageEvent {
    data class Selection(val text: String) : ClipPageEvent()
    data class TagTapped(val field: ClipField) : ClipPageEvent()
    data class ImageTapped(val src: String) : ClipPageEvent()
}

/** How the page is loaded: the live URL, or a fixed local page in a UI test. */
typealias ClipPageLoader = (WebView, String) -> Unit

internal val LoadLiveUrl: ClipPageLoader = { webView, url -> webView.loadUrl(url) }

/**
 * The page being clipped, in a [WebView] with `clipper.js` injected once it has loaded. The
 * view layer's half of the bridge: it forwards the page's events and pushes [syncState] and
 * [pickingPhoto] into the page whenever they change. Links to other pages are blocked, so the
 * clip always comes from the page it is saved under.
 */
@SuppressLint("SetJavaScriptEnabled", "JavascriptInterface")
@Composable
internal fun ClipWebPage(
    url: String,
    syncState: String,
    pickingPhoto: Boolean,
    onEvent: (ClipPageEvent) -> Unit,
    loadPage: ClipPageLoader,
    modifier: Modifier = Modifier
) {
    // Read by the WebViewClient when a page finishes loading, so the script it injects starts
    // from the current state rather than the state at creation.
    val latest = remember { LatestPageState() }
    latest.sync = syncState
    latest.picking = pickingPhoto
    latest.onEvent = onEvent

    AndroidView(
        modifier = modifier,
        factory = { context ->
            WebView(context).apply {
                settings.javaScriptEnabled = true
                settings.domStorageEnabled = true
                val main = Handler(Looper.getMainLooper())
                addJavascriptInterface(
                    object {
                        // Called on the JavaBridge thread; the ViewModel is fed on the main one.
                        @JavascriptInterface
                        fun post(json: String) {
                            val event = decode(json) ?: return
                            main.post { latest.onEvent(event) }
                        }
                    },
                    "RCAndroid"
                )
                webViewClient = ClipWebViewClient(url, latest)
                loadPage(this, url)
            }
        },
        // Captures the two values, so it is a new lambda, and runs again, when either changes.
        update = { webView ->
            if (latest.injected) webView.evaluateJavascript(pageScript(syncState, pickingPhoto), null)
        }
    )
}

private class LatestPageState {
    var sync: String = "{}"
    var picking: Boolean = false
    var onEvent: (ClipPageEvent) -> Unit = {}
    var injected: Boolean = false

    fun script() = pageScript(sync, picking)
}

private fun pageScript(sync: String, picking: Boolean) =
    "window.RC && (RC.sync($sync), RC.pickImage($picking));"

private class ClipWebViewClient(
    private val pageUrl: String,
    private val latest: LatestPageState
) : WebViewClient() {

    override fun onPageStarted(view: WebView, url: String?, favicon: android.graphics.Bitmap?) {
        latest.injected = false
    }

    override fun onPageFinished(view: WebView, url: String?) {
        view.evaluateJavascript(ClipperScript.source, null)
        latest.injected = true
        view.evaluateJavascript(latest.script(), null)
    }

    /** Redirects and fragment jumps load; a link to any other page does not. */
    override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
        if (!request.isForMainFrame || request.isRedirect) return false
        return !samePage(request.url, (view.url ?: pageUrl).toUri())
    }

    private fun samePage(a: Uri, b: Uri): Boolean =
        a.buildUpon().fragment(null).build() == b.buildUpon().fragment(null).build()
}

/** `shared/web/clipper.js`, packaged as a Java resource alongside the shared tables. */
internal object ClipperScript {
    val source: String by lazy {
        ClipperScript::class.java.getResourceAsStream("/web/clipper.js")!!
            .bufferedReader().use { it.readText() }
    }
}

private fun decode(json: String): ClipPageEvent? = try {
    val message = JSONObject(json)
    when (message.optString("type")) {
        "selection" -> ClipPageEvent.Selection(message.optString("text"))
        "tag" -> ClipField.entries.firstOrNull { it.name == message.optString("field") }
            ?.let { ClipPageEvent.TagTapped(it) }
        "image" -> message.optString("src").takeIf { it.isNotEmpty() }?.let { ClipPageEvent.ImageTapped(it) }
        else -> null
    }
} catch (e: org.json.JSONException) {
    null
}
